File: bridge/rdp-install.pl
#!/usr/bin/perl
# Install the extension into Firefox via the Remote Debugging Protocol (RDP) -
# the same channel web-ext uses. Unlike Marionette, RDP does NOT set the
# navigator.webdriver "automation" flag, so sites won't throw robot-checks.
# Firefox must be launched with:  -start-debugger-server <port>
# and the profile must allow it (installer writes the prefs into user.js).
#
# Usage: perl rdp-install.pl [port] [path-to-extension-dir]
use strict;
use warnings;
use IO::Socket::INET;
use JSON::PP;
use File::Basename;
use Cwd qw(abs_path);

my $PORT = $ARGV[0] || 6000;
my $EXT  = $ARGV[1] || (dirname(dirname(abs_path($0))) . "/extension");
my $json = JSON::PP->new->utf8;

my $sock;
for (my $i = 0; $i < 80; $i++) {
    $sock = IO::Socket::INET->new(PeerAddr => '127.0.0.1', PeerPort => $PORT, Proto => 'tcp');
    last if $sock;
    select(undef, undef, undef, 0.5);
}
die "ERROR: could not reach Firefox debugger on 127.0.0.1:$PORT\n" unless $sock;
$sock->autoflush(1);

# RDP framing: <byte-length>:<json-payload>
sub read_msg {
    my ($s) = @_;
    my $lenstr = '';
    while (1) {
        my $c; my $n = sysread($s, $c, 1);
        die "ERROR: connection closed reading length\n" unless $n;
        last if $c eq ':';
        $lenstr .= $c;
    }
    my $len = $lenstr + 0;
    my $buf = '';
    while (length($buf) < $len) {
        my $chunk; my $n = sysread($s, $chunk, $len - length($buf));
        die "ERROR: connection closed reading body\n" unless $n;
        $buf .= $chunk;
    }
    return $json->decode($buf);
}
sub send_msg {
    my ($s, $obj) = @_;
    my $p = $json->encode($obj);
    print $s length($p) . ':' . $p;
}

my $hello = read_msg($sock);
print "GREETING: applicationType=" . ($hello->{applicationType} // '?') . "\n";

send_msg($sock, { to => 'root', type => 'getRoot' });
my $root = read_msg($sock);
my $addonsActor = $root->{addonsActor};
die "ERROR: no addonsActor in root (RDP not enabled?)\n" unless $addonsActor;

send_msg($sock, { to => $addonsActor, type => 'installTemporaryAddon', addonPath => $EXT });
my $res = read_msg($sock);
die "ERROR installTemporaryAddon: " . $json->encode($res) . "\n" if $res->{error};
my $id = ref($res->{addon}) eq 'HASH' ? ($res->{addon}{id} // '?') : ($res->{id} // '?');
print "INSTALLED: $id\n";
close($sock);

File: server/rdp-install.js
'use strict';
/**
 * Load the bridge extension into Firefox over the Remote Debugging Protocol.
 *
 * WHY RDP and not Marionette: RDP is the same channel web-ext uses and it does NOT set
 * navigator.webdriver, so websites never throw robot-checks at the user. Marionette does.
 *
 * WHY THIS MATTERS FOR SHARING: an add-on loaded this way is a TEMPORARY add-on, which
 * Firefox permits unsigned. Nobody needs a Mozilla developer account, AMO signing or 2FA
 * to use this connector - it simply loads each time the bridge starts.
 *
 * Firefox must be running with -start-debugger-server <port>, on a profile whose user.js
 * enables remote debugging (the launcher writes exactly that).
 */
const net = require('net');

const ADDON_ID = 'claude-firefox-bridge@bentaco.local';

/** RDP framing is <byte-length>:<json-payload>. */
function makeReader(sock) {
  let buf = Buffer.alloc(0);
  const waiters = [];

  sock.on('data', (chunk) => {
    buf = Buffer.concat([buf, chunk]);
    for (;;) {
      const colon = buf.indexOf(0x3a); // ':'
      if (colon < 0) return;
      const len = parseInt(buf.slice(0, colon).toString('ascii'), 10);
      if (!Number.isFinite(len)) return;
      if (buf.length < colon + 1 + len) return;
      const body = buf.slice(colon + 1, colon + 1 + len).toString('utf8');
      buf = buf.slice(colon + 1 + len);
      let msg;
      try { msg = JSON.parse(body); } catch (e) { msg = {}; }
      const w = waiters.shift();
      if (w) w(msg);
    }
  });

  return function read(timeoutMs = 20000) {
    return new Promise((resolve, reject) => {
      const t = setTimeout(() => reject(new Error('timed out reading from the Firefox debugger')),
                           timeoutMs);
      waiters.push((m) => { clearTimeout(t); resolve(m); });
    });
  };
}

function send(sock, obj) {
  const p = Buffer.from(JSON.stringify(obj), 'utf8');
  sock.write(Buffer.from(String(p.length) + ':', 'ascii'));
  sock.write(p);
}

function connect(port, timeoutMs = 40000) {
  const deadline = Date.now() + timeoutMs;
  return new Promise((resolve) => {
    (function attempt() {
      const s = net.connect({ host: '127.0.0.1', port }, () => resolve(s));
      s.on('error', () => {
        s.destroy();
        if (Date.now() > deadline) return resolve(null);
        setTimeout(attempt, 500);
      });
    })();
  });
}

/** Install the extension. Resolves {ok, message}. */
async function install(port, extPath) {
  const sock = await connect(port);
  if (!sock) {
    return { ok: false, message: 'could not reach the Firefox debugger on 127.0.0.1:' + port };
  }
  try {
    const read = makeReader(sock);
    await read();                                   // root greeting

    send(sock, { to: 'root', type: 'getRoot' });
    const root = await read();
    const addons = root.addonsActor;
    if (!addons) return { ok: false, message: 'no addonsActor (remote debugging not enabled?)' };

    // Uninstall first to FLUSH the StartupCache bytecode for this fixed add-on id.
    // Temporary RDP installs cache compiled code by id; without this unmount an edited
    // background.js does NOT reload - the old bytecode keeps being served. On a fresh
    // profile it isn't installed yet, so errors here are expected and ignored.
    send(sock, { to: addons, type: 'uninstallTemporaryAddon', addonId: ADDON_ID });
    try { await read(4000); } catch (e) {}

    send(sock, { to: addons, type: 'installTemporaryAddon', addonPath: extPath });
    const res = await read();
    if (res.error) return { ok: false, message: 'installTemporaryAddon failed: ' + JSON.stringify(res) };
    const id = (res.addon && res.addon.id) || res.id || '?';
    return { ok: true, message: 'installed ' + id };
  } catch (e) {
    return { ok: false, message: 'RDP error: ' + e.message };
  } finally {
    try { sock.destroy(); } catch (e) {}
  }
}

module.exports = { install, ADDON_ID };

if (require.main === module) {
  const port = parseInt(process.argv[2] || '6000', 10);
  const ext = process.argv[3] || '';
  install(port, ext).then((r) => {
    console.log((r.ok ? 'OK: ' : 'ERROR: ') + r.message);
    process.exit(r.ok ? 0 : 1);
  });
}
